plugins/com.google.cloud.tools.eclipse.dataflow.core/src/com/google/cloud/tools/eclipse/dataflow/core/project/DataflowArtifactRetriever.java

```diff
@@ -0,0 +1,227 @@
+/*
+ * Copyright 2015 Google Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package com.google.cloud.tools.eclipse.dataflow.core.project;
+
+import static com.google.common.base.Preconditions.checkArgument;
+
+import com.google.cloud.tools.eclipse.dataflow.core.DataflowCorePlugin;
+import com.google.cloud.tools.eclipse.dataflow.core.project.DataflowProjectCreator.Template;
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.cache.CacheBuilder;
+import com.google.common.cache.CacheLoader;
+import com.google.common.cache.LoadingCache;
+import com.google.common.collect.ImmutableSortedSet;
+import com.google.common.collect.ImmutableSortedSet.Builder;
+import org.apache.maven.artifact.versioning.ArtifactVersion;
+import org.apache.maven.artifact.versioning.DefaultArtifactVersion;
+import org.apache.maven.artifact.versioning.VersionRange;
+import org.eclipse.core.runtime.CoreException;
+import org.eclipse.core.runtime.Status;
+import org.w3c.dom.Document;
+import org.w3c.dom.NodeList;
+import org.xml.sax.SAXException;
+import java.io.IOException;
+import java.net.MalformedURLException;
+import java.net.URL;
+import java.util.NavigableSet;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.TimeUnit;
+import javax.xml.parsers.DocumentBuilderFactory;
+import javax.xml.parsers.ParserConfigurationException;
+import javax.xml.xpath.XPath;
+import javax.xml.xpath.XPathConstants;
+import javax.xml.xpath.XPathExpression;
+import javax.xml.xpath.XPathExpressionException;
+import javax.xml.xpath.XPathFactory;
+
+/**
+ * {@link DataflowArtifactRetriever} provides access to Maven artifacts for Dataflow projects, as
+ * well as the examples source code.
+ *
+ * <p>The implementation of {@link DataflowArtifactRetriever} uses low-level URL and XPath APIs
+ * rather than using the M2E plugin to work around shortcomings in the ability of M2E to query
+ * Maven for available versions. The artifact retriever reads Maven Central metadata xml files to
+ * retrieve available and latest versions.
+ */
+public class DataflowArtifactRetriever {
```

Comment:
We've been inventing something very much like this though it looks like you're a little further along. We're definitely going to want to sync this at some point.